File: Leetcode/easy/ValidAnagram.scala
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.
// Time O(n), Space O(n)

object Solution {
    def isAnagram(s: String, t: String): Boolean = {
        def charMap(string: String) = string.groupBy(identity).view.mapValues(_.length).toMap
        charMap(s) == charMap(t)
    }
}

File: Leetcode/easy/ValidAnagram.js
// Given two strings s and t, return true if t is an anagram of s, and false otherwise.
// Time O(n), Space O(n)

const isAnagram = function (s, t) {
    if (s.length !== t.length) return false;

    const counts = {};
    for (let c of s) {
        counts[c] = (counts[c] || 0) + 1;
    }

    for (let c of t) {
        if (counts[c]) counts[c]--;
        else return false;
    }

    return true;
};